libp2p/connectionsHandler.go

```diff
@@ -0,0 +1,389 @@
+	ctxMaxWaitingTime, cancel := context.WithTimeout(context.Background(), maxWaitingTime)
+	defer cancel()
+
+	for {
+		if handler.shouldStopWaiting(ctxMaxWaitingTime, minNumOfPeers) {
+			return
+		}
+	}
+}
+
+func (handler *connectionsHandler) shouldStopWaiting(ctxMaxWaitingTime context.Context, minNumOfPeers uint32) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), pollWaitForConnectionsInterval)
+	defer cancel()
+
+	select {
+	case <-ctxMaxWaitingTime.Done():
+		return true
+	case <-ctx.Done():
+		return int(minNumOfPeers) <= len(handler.ConnectedPeers())
+	}
+}
+
+// IsConnected returns true if current node is connected to provided peer
+func (handler *connectionsHandler) IsConnected(peerID core.PeerID) bool {
+	h := handler.p2pHost
+
+	connectedness := h.Network().Connectedness(peer.ID(peerID))
+
+	return connectedness == network.Connected
+}
+
+// ConnectedPeers returns the current connected peers list
+func (handler *connectionsHandler) ConnectedPeers() []core.PeerID {
+	h := handler.p2pHost
+
+	connectedPeers := make(map[core.PeerID]struct{})
+
+	for _, conn := range h.Network().Conns() {
+		p := core.PeerID(conn.RemotePeer())
+
+		if handler.IsConnected(p) {
+			connectedPeers[p] = struct{}{}
+		}
+	}
+
+	peerList := make([]core.PeerID, len(connectedPeers))
+
+	index := 0
+	for k := range connectedPeers {
+		peerList[index] = k
+		index++
+	}
+
+	return peerList
+}
+
+// ConnectedAddresses returns all connected peer's addresses
+func (handler *connectionsHandler) ConnectedAddresses() []string {
+	h := handler.p2pHost
+	conns := make([]string, 0)
+
+	for _, c := range h.Network().Conns() {
+		conns = append(conns, c.RemoteMultiaddr().String()+"/p2p/"+c.RemotePeer().String())
+	}
+	return conns
+}
+
+// PeerAddresses returns the peer's addresses or empty slice if the peer is unknown
+func (handler *connectionsHandler) PeerAddresses(pid core.PeerID) []string {
+	h := handler.p2pHost
+	result := make([]string, 0)
+
+	// check if the peer is connected to the node and append its address
+	for _, c := range h.Network().Conns() {
+		if string(c.RemotePeer()) == string(pid.Bytes()) {
+			result = append(result, c.RemoteMultiaddr().String())
+			break
+		}
+	}
+
+	// check in peerstore (maybe it is known but not connected)
+	addresses := h.Peerstore().Addrs(peer.ID(pid.Bytes()))
+	for _, addr := range addresses {
+		result = append(result, addr.String())
+	}
+
+	return result
+}
+
+// ConnectedPeersOnTopic returns the connected peers on a provided topic
+func (handler *connectionsHandler) ConnectedPeersOnTopic(topic string) []core.PeerID {
+	return handler.peersOnChannel.ConnectedPeersOnChannel(topic)
+}
+
+// ConnectedFullHistoryPeersOnTopic returns the connected peers on a provided topic
+func (handler *connectionsHandler) ConnectedFullHistoryPeersOnTopic(topic string) []core.PeerID {
+	peerList := handler.ConnectedPeersOnTopic(topic)
+	fullHistoryList := make([]core.PeerID, 0)
+
+	handler.mutPeerResolver.RLock()
```

Comment:
Good find here for the missing lock. However, I would keep it only to get the `peerShardResolver` pointer